src/walinfo.c

```diff
@@ -92,6 +94,253 @@ usage(void)
    printf("Report bugs: %s\n", PGMONETA_ISSUES);
 }
 
+static int compare_walfile(struct walfile* wf1, struct walfile* wf2);
+static bool compare_long_page_headers(struct xlog_long_page_header_data* h1, struct xlog_long_page_header_data* h2);
+static bool compare_deque(struct deque* dq1, struct deque* dq2, bool (*compare)(void*, void*));
+static bool compare_xlog_page_header(void* a, void* b);
+static bool compare_xlog_record(void* a, void* b);
+
+static int
+compare_walfile(struct walfile* wf1, struct walfile* wf2)
+{
+   // Handle NULL cases
+   if (wf1 == NULL || wf2 == NULL)
+   {
+      return (wf1 == wf2) ? 0 : -1;
+   }
+
+   // Compare magic number (already in your code)
+   // if (wf1->magic_number != wf2->magic_number)
+   // {
+   //    printf("Magic number mismatch: %u != %u\n", wf1->magic_number, wf2->magic_number);
+   //    return -1;
+   // }
+
+   // Compare long page headers (long_phd)
+   if (!compare_long_page_headers(wf1->long_phd, wf2->long_phd))
+   {
+      printf("Long page header mismatch\n");
+      return -1;
+   }
+
+   // Compare page headers deque
+   if (!compare_deque(wf1->page_headers, wf2->page_headers, compare_xlog_page_header))
+   {
+      printf("Page headers deque mismatch\n");
+      return -1;
+   }
+
+   // Compare records deque
+   if (!compare_deque(wf1->records, wf2->records, compare_xlog_record))
+   {
+      printf("Records deque mismatch\n");
+      return -1;
+   }
+
+   return 0;
+}
+
+// Helper: Compare XLogLongPageHeaderData
+static bool
+compare_long_page_headers(struct xlog_long_page_header_data* h1, struct xlog_long_page_header_data* h2)
+{
+   if (h1 == NULL || h2 == NULL)
+   {
+      return (h1 == h2);
+   }
+
+   return (h1->std.xlp_magic == h2->std.xlp_magic &&
+           h1->std.xlp_info == h2->std.xlp_info &&
+           h1->std.xlp_tli == h2->std.xlp_tli &&
+           h1->std.xlp_pageaddr == h2->std.xlp_pageaddr &&
+           h1->xlp_seg_size == h2->xlp_seg_size &&
+           h1->xlp_xlog_blcksz == h2->xlp_xlog_blcksz);
+}
+
+// Helper: Generic deque comparison with custom comparator
+static bool
+compare_deque(struct deque* dq1, struct deque* dq2, bool (*compare)(void*, void*))
```

Comment:
This belongs in `deque.h`

Reply:
I don't think we need a generic deque comparator. This looks like something very specific for walinfo at the moment.